File: mirea/cources/2026-2027/aii-july/02/Part-04-Monostack/examples/example-02.py
numbers = [5, 2, 4, 1, 3]
stack = []
next_smaller = [-1] * len(numbers)

# Идём справа налево и ищем ближайший меньший элемент справа.
for i in range(len(numbers) - 1, -1, -1):
    while stack and numbers[stack[-1]] >= numbers[i]:
        stack.pop()

    if stack:
        next_smaller[i] = numbers[stack[-1]]

    stack.append(i)

print("next smaller values:", *next_smaller)

File: mirea/cources/2026-2027/aii-july/02/Part-04-Monostack/examples/example-02.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    vector<int> a = {5, 2, 4, 1, 3};
    vector<int> st;
    vector<int> next_smaller(a.size(), -1);

    // Идём справа налево и ищем ближайший меньший элемент справа.
    for (int i = (int)a.size() - 1; i >= 0; i--) {
        while (!st.empty() && a[st.back()] >= a[i]) {
            st.pop_back();
        }

        if (!st.empty()) {
            next_smaller[i] = a[st.back()];
        }

        st.push_back(i);
    }

    cout << "next smaller values: ";
    for (int value : next_smaller) {
        cout << value << ' ';
    }
    cout << '\n';

    return 0;
}